### PR TITLE
new bi sample benchmark tests v2 (fromPR #1995)

### PR DESCRIPTION
**Pull request by grusev**
#### Reference Issues/PRs

From PR #1995 (all comments addressed)

    Sample test benchmark for using one opensource BI CSV source.
    The logic of a test is 
        - download if parquet file does not exists source in .bz2 format
        - convert it to parquet format
        - prepare library with it containing  several symbols that are constructed based on this DF
        - for each query we want to benchmark do a pre-check that this query produces SAME result on Pandas and arcticDB
        - run the benchmark tests

#### Checklist

<details>
  <summary>
   Checklist for code changes...
  </summary>
 
 - [ ] Have you updated the relevant docstrings, documentation and copyright notice?
 - [ ] Is this contribution tested against [all ArcticDB's features](../docs/mkdocs/docs/technical/contributing.md)?
 - [ ] Do all exceptions introduced raise appropriate [error messages](https://docs.arcticdb.io/error_messages/)?
 - [ ] Are API changes highlighted in the PR description?
 - [ ] Is the PR labelled as enhancement or bug so it appears in autogenerated release notes?
</details>

### CODE REVIEW

File: python/data/CityMaxCapita_1.parquet.gzip
Comment (alexowens90) [inline python/data/CityMaxCapita_1.parquet.gzip:1]: parquet is already a compressed format, so there probably isn't any additional benefit to gzip

File: python/benchmarks/bi_benchmarks.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+        print("The procedure is creating N times larger dataframes")
+        print("by concatenating original DF N times")
+        for num in BIBenchmarks.params:
+            _df = self.df.copy(deep=True)
+            if (num > 1):
+                # lets create N times bigger DF
+                dfcum = self.df.copy(deep=True)
+                for i in range(1, (BIBenchmarks.params[-1])):
+                    dfcum = pd.concat([dfcum, self.df])
```
Comment (alexowens90) [inline python/benchmarks/bi_benchmarks.py:89]: This is much less efficient that calling concat once with a longer list of dataframes

File: python/benchmarks/bi_benchmarks.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+    number = 2
+    timeout = 6000
+    LIB_NAME = "BI_benchmark_lib"
+    # We use dataframe in this file
+    CITY_BI_FILE = "data/CityMaxCapita_1.csv.bz2"
+    CITY_BI_FILE2 = "data/CityMaxCapita_1.parquet.gzip"
+
+    #Defines how many times bigger the database is
+    params = [1, 10]
```
Comment (alexowens90) [inline python/benchmarks/bi_benchmarks.py:55]: How big is the original data set in terms of number of rows? How long does this whole thing take to run?

File: python/benchmarks/bi_benchmarks.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+import os
+from pathlib import Path
+from arcticdb import Arctic
+from arcticdb.version_store.processing import QueryBuilder
+
+from .common import *
+
+
+def get_query_groupby_city_count_all(q):
+    return q.groupby("City").agg({"Keyword": "count"})
```
Comment (alexowens90) [inline python/benchmarks/bi_benchmarks.py:17]: nit: it would be better programming practice to construct the `QueryBuilder` in these methods and then return it, rather than taking a parameter.

File: python/benchmarks/bi_benchmarks.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    params = [1, 10]
+
+    def __init__(self):
+        self.lib_name = BIBenchmarks.LIB_NAME
+        self.symbol = self.lib_name
+
+    def setup_cache(self):
+
+        file = os.path.join(Path(__file__).resolve().parent.parent, BIBenchmarks.CITY_BI_FILE2)
+        if (not os.path.exists(file)) :
```
Comment (alexowens90) [inline python/benchmarks/bi_benchmarks.py:64]: Why is this needed since the file is checked in?

File: python/data/CityMaxCapita_1.parquet.gzip
Comment (grusev) [reply python/data/CityMaxCapita_1.parquet.gzip:1]: size of parquet file is 800MB when gzipped it is 95MB

File: python/benchmarks/bi_benchmarks.py (hunk 3 above)
Comment (grusev) [reply python/benchmarks/bi_benchmarks.py:17]: Agree, but in this particular case I use those methods to pass QueryBuilder or DataFrame, so that I can use same methods to work with arctic and dataframes (pandas (see the setup cache)

Comment (grusev) [reply python/benchmarks/bi_benchmarks.py:17]: But you have point as this si not obvious, I will add type hints here

File: python/benchmarks/bi_benchmarks.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+def get_query_groupby_city_count_filter_two_aggregations(q):
+    return q[q["Keyword"] == "faggot" ].groupby("City").agg({"Keyword": "count", "Number of Records" : "sum"})  
+
+def assert_frame_equal(pandas_df:pd.DataFrame, arctic_df:pd.DataFrame):
+    arctic_df.sort_index(inplace=True)
+    pd.testing.assert_frame_equal(pandas_df,
+                                    arctic_df, 
+                                    check_column_type=False, 
+                                    check_dtype=False)
```
[the remark this answers is not in the document]
Comment (grusev) [reply python/benchmarks/bi_benchmarks.py:32]: The one that we have is not suitable for this and I had to come up with one that will not fail due to not relevant things

File: python/benchmarks/bi_benchmarks.py (hunk 2 above)
Comment (grusev) [reply python/benchmarks/bi_benchmarks.py:55]: 912k rows. 

25 secs generating and doing functional checks over the original and large 10x dataframe (9120k rows)

See more details for the test:

OUTPUT -------->
             Parquet file exists!
             The procedure is creating N times larger dataframes
             by concatenating original DF N times
             Size of original Dataframe:  912657
             DF for iterration xSize original ready:  10
             <class 'pandas.core.frame.DataFrame'>
             Index: 9126570 entries, 0 to 912656
             Data columns (total 31 columns):
              #   Column               Dtype
             ---  ------               -----
              0   City/Admin           object
              1   City/State           object
              2   City                 object
              3   Created Date/Time    float64
              4   Date Joined          float64
              5   FF Ratio             float64
              6   Favorites            int32
              7   First Link in Tweet  object
              8   Followers            int32
              9   Following            int32
              10  Gender               object
              11  Influencer?          int32
              12  Keyword              object
              13  LPF                  float64
              14  Language             object
              15  Lat                  float64
              16  Listed Number        int32
              17  Long Domain          object
              18  Long                 float64
              19  Number of Records    int32
              20  Region               object
              21  Short Domain         object
              22  State/Country        object
              23  State                object
              24  Tweet Text           object
              25  Tweets               int32
              26  Twitter Client       object
              27  User Bio             object
              28  User Loc             object
              29  Username 1           object
              30  Username             object
             dtypes: float64(6), int32(7), object(18)
             memory usage: 10.8 GB
             If pandas query produces different dataframe than arctic one stop tests!
             This will mean query problem is there most likely
             Pre-check correctness for query_groupby_city_count_all
             Pre-check correctness for query_groupby_city_count_isin_filter
             Pre-check correctness for query_groupby_city_count_filter_two_aggregations
             All pre-checks completed SUCCESSFULLY. Time:  25.181268453598022
                                                                            ok
[31.25%] ··· Running (bi_benchmarks.BIBenchmarks.time_query_groupby_city_count_all--)....
[56.25%] ··· bi_benchmarks.BIBenchmarks.peakmem_query_groupby_city_count_all                                        ok
[56.25%] ··· ======== =======
              param1
             -------- -------
                1       372M
                10     2.77G
             ======== =======

[62.50%] ··· bi_benchmarks.BIBenchmarks.peakmem_query_groupby_city_count_filter_two_aggregations                    ok
[62.50%] ··· ======== =======
              param1
             -------- -------
                1       363M
                10     2.73G
             ======== =======

[68.75%] ··· bi_benchmarks.BIBenchmarks.peakmem_query_groupby_city_count_isin_filter                                ok
[68.75%] ··· ======== =======
              param1
             -------- -------
                1       361M
                10     2.69G
             ======== =======

[75.00%] ··· bi_benchmarks.BIBenchmarks.peakmem_query_readall                                                       ok
[75.00%] ··· ======== =======
              param1
             -------- -------
                1       936M
                10     8.06G
             ======== =======

[81.25%] ··· bi_benchmarks.BIBenchmarks.time_query_groupby_city_count_all                                           ok
[81.25%] ··· ======== ==========
              param1
             -------- ----------
                1      19.8±3ms
                10     163±4ms
             ======== ==========

[87.50%] ··· bi_benchmarks.BIBenchmarks.time_query_groupby_city_count_filter_two_aggregations                       ok
[87.50%] ··· ======== ==========
              param1
             -------- ----------
                1      21.3±3ms
                10     155±4ms
             ======== ==========

[93.75%] ··· bi_benchmarks.BIBenchmarks.time_query_groupby_city_count_isin_filter                                   ok
[93.75%] ··· ======== ==========
              param1
             -------- ----------
                1      20.0±2ms
                10     154±3ms
             ======== ==========

[100.00%] ··· bi_benchmarks.BIBenchmarks.time_query_readall                                                          ok
[100.00%] ··· ======== ============
               param1
              -------- ------------
                 1       365±20ms
                 10     3.54±0.02s
              ======== ============

File: python/benchmarks/bi_benchmarks.py (hunk 4 above)
Comment (grusev) [reply python/benchmarks/bi_benchmarks.py:64]: Originally I implemented only downloading, but it was slow. Then decided to leave this here, so that in future thise code can be copied and run even without having the data with you

File: python/benchmarks/bi_benchmarks.py (hunk 1 above)
Comment (grusev) [reply python/benchmarks/bi_benchmarks.py]: True!!!

File: python/benchmarks/bi_benchmarks.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+            _df = self.df.copy(deep=True)
+            if (num > 1):
+                # lets create N times bigger DF
+                dfcum = self.df.copy(deep=True)
+                for i in range(1, (BIBenchmarks.params[-1])):
+                    dfcum = pd.concat([dfcum, self.df])            
+                _df = dfcum
+                print("DF for iterration xSize original ready: ", num)
+                _df.info(verbose=True,memory_usage='deep')
+            self.lib.write(f"{self.symbol}{num}", _df)
```
[the remark this answers is not in the document]
Comment (alexowens90) [reply python/benchmarks/bi_benchmarks.py:104]: Ah OK. In that case this can still be tidied up, the copying is unnecessary:
```
for num in BIBenchmarks.params:
    _df = pd.concat([self.df] * num)
    print("DF for iterration xSize original ready: ", num)
    self.lib.write(f"{self.symbol}{num}", _df)
```

File: python/benchmarks/bi_benchmarks.py (hunk 2 above)
Comment (alexowens90) [reply python/benchmarks/bi_benchmarks.py:59]: And how about the runtime including the actual benchmarking? ASV runs each benchmark multiple times to reduce the standard deviation of the output, so it isn't clear what the total runtime would be from these numbers

File: python/benchmarks/bi_benchmarks.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexowens90) [reply python/benchmarks/bi_benchmarks.py:35]: It just forwards to `pd.testing.assert_frame_equal`, so how is it not suitable?

File: python/benchmarks/bi_benchmarks.py (hunk 3 above)
Comment (alexowens90) [reply python/benchmarks/bi_benchmarks.py:19]: Ah I see, makes sense

File: python/benchmarks/common.py
Comment (alexowens90) [inline python/benchmarks/common.py:1]: Any reason these changes are in `common.py`? They're only used by `bi_benchmarks.py`, so should live in there, no?

File: python/benchmarks/bi_benchmarks.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (grusev) [reply python/benchmarks/bi_benchmarks.py:104]: done

File: python/benchmarks/bi_benchmarks.py (hunk 2 above)
Comment (grusev) [reply python/benchmarks/bi_benchmarks.py:59]: total setup + tests <50secs

File: python/benchmarks/bi_benchmarks.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (grusev) [reply python/benchmarks/bi_benchmarks.py]: done
